lib/documents.js

```diff
@@ -2263,12 +2264,185 @@ function writeAllDocumentsImpl(options) {
+
+   //Check for defaultMetadata
+   jobState.jobOptions.defaultMetadata = (jobState.jobOptions.defaultMetadata)? jobState.jobOptions.defaultMetadata:null;
+
+   //Check for transform
+   jobState.jobOptions.transform = (jobState.jobOptions.transform)? jobState.jobOptions.transform:null;
+
+  requester.startRequest(operation).result(mlutil.callbackOn(jobState, onWriteAllInit))
+      .catch((err) => operation.stream.emit('error', err));
+
   return inputStream;
 }
 
+function finishWriter(jobState) {
+  jobState.requesterCount--;
+  if(jobState.requesterCount === 0) {
+    if(jobState.jobOptions && jobState.jobOptions.onCompletion){
+      const summary = {
+        docsWrittenSuccessfully: jobState.docsWrittenSuccessfully,
+        docsFailedToBeWritten: jobState.docsFailedToBeWritten,
+        timeElapsed: jobState.timeElapsed
+      };
+      if (jobState.error) {
+        summary.error = jobState.error.toString();
+      }
+      jobState.jobOptions.onCompletion(summary);
+    }
+  }
+}
+
+function onWriteAllInit(output) {
+
+  let maxRequesters = 0;
+  const jobState = this;
+
+  switch (jobState.jobOptions.concurrentRequests.multipleOf) {
+    case 'forests':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier * output.length;
+      break;
+    case 'hosts':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier *
+          new Set(output.map((forest) => forest.host)).size;
+      break;
+    default: {
+      const err =  new Error('Invalid value for onCompletion.multipleOf. Value must be forests or hosts.');
+      jobState.error = err;
+      jobState.stream.emit('error', err);
+    }
+  }
+  jobState.startTime = Date.now();
+  for (let i = 0; i < maxRequesters; i++) {
+    jobState.requesterCount++;
+    onWriteAllDocs(jobState,i);
+  }
+}
+
+function onWriteAllDocs(jobState, writerId) {
+
+  const writeBatchArray = [];
+  const batchSize = jobState.jobOptions.batchSize;
+  jobState.lastBatch = false;
+  jobState.retryCount = 2;
+
+  a:
+      for (let i = 0; i < batchSize; i++) {
+        const record = jobState.stream.read();
+
+        if(record !== null) {
+          if (jobState.jobOptions.defaultMetadata && i === 0) {
+            writeBatchArray.push(jobState.jobOptions.defaultMetadata);
+          }
+        }
+        else {
+          switch (writeBatchArray.length) {
+            case 0: {
+              jobState.timeElapsed = Date.now() - jobState.startTime;
+              finishWriter(jobState);
+              break a;
+            }
+            default: {
+              jobState.lastBatch = true;
+              break a;
+            }
+          }
+        }
+        writeBatchArray.push(record);
+      }
+
+  if (writeBatchArray.length !== 0) {
+    const batchdef = [];
+    batchdef.push({documents: writeBatchArray});
+    if (jobState.jobOptions.transform) {
+      batchdef.push({transform:jobState.jobOptions.transform});
```

Comment:
I think a comment on a prior review iteration got lost in the revisions.  I probably resolved a comment prematurely.

The transform isn't pushed on the batch definition.  It's a parameter of the call to writeDocumentsImpl().
See the JSDoc in the source file before the writeDocuments() definition for the order of the parameters.

By the way, it will be important for a unit test to confirm that the default metadata (if supplied) is written.